README.md
# `flutter_bluetooth_serial`

[![pub package](https://img.shields.io/pub/v/flutter_bluetooth_serial.svg)](https://pub.dartlang.org/packages/flutter_bluetooth_serial)

Flutter basic implementation for Classical Bluetooth (only RFCOMM for now).

## Features

The first goal of this project, started by @edufolly was making an interface for
Serial Port Protocol (HC-05 Adapter). Now the plugin features:

+ Adapter status monitoring,

+ Turning adapter on and off,

+ Opening settings,

+ Discovering devices (and requesting discoverability),

+ Listing bonded devices and pairing new ones,

+ Connecting to multiple devices at the same time,

+ Sending and receiving data (multiple connections).

The plugin (for now) uses Serial Port profile for moving data over RFCOMM, so
make sure there is running Service Discovery Protocol that points to SP/RFCOMM
channel of the device. There could
be [max up to 7 Bluetooth connections](https://stackoverflow.com/a/32149519/4880243).

For now there is only Android support.

## Funding

Your contribution will help drive the development of quality tools for the
Flutter and Dart developer community. Any amount will be appreciated. Thank you
for your continued support!

[![BuyMeACoffee](https://www.buymeacoffee.com/assets/img/guidelines/download-assets-sm-2.svg)](https://www.buymeacoffee.com/edufolly)

## PIX

Sua contribuição ajudará a impulsionar o desenvolvimento de ferramentas de
qualidade para a comunidade de desenvolvedores Flutter e Dart. Qualquer quantia
será apreciada. Obrigado pelo seu apoio contínuo!

[![PIX](helpers/pix.png)](https://nubank.com.br/pagar/2bt2q/RBr4Szfuwr)

## Getting Started

#### Depending

```yaml
# Add dependency to `pubspec.yaml` of your project.
dependencies:
  # ...
  flutter_bluetooth_serial: ^0.4.0
```

#### Installing

```bash
flutter pub get
```

#### Importing

```dart
import 'package:flutter_bluetooth_serial/flutter_bluetooth_serial.dart';
```

#### Usage

You should look to the Dart code of the library (mostly documented functions) or
to the examples code.

```dart
// Some simplest connection :F
try {
    BluetoothConnection connection = await BluetoothConnection.toAddress(address);
    print('Connected to the device');

    connection.input.listen((Uint8List data) {
        print('Data incoming: ${ascii.decode(data)}');
        connection.output.add(data); // Sending data

        if (ascii.decode(data).contains('!')) {
            connection.finish(); // Closing connection
            print('Disconnecting by local host');
        }
    }).onDone(() {
        print('Disconnected by remote request');
    });
}
catch (exception) {
    print('Cannot connect, exception occured');
}
```

Note: Work is underway to make the communication easier than operations on byte
streams. See #41 for discussion about the topic.

#### Examples

Check out [example application](example/README.md) with connections with both
Arduino HC-05 and Raspberry Pi (RFCOMM) Bluetooth interfaces.

|       Main screen and options        |       Discovery and connecting       |       Simple chat with server        |        Background connection         |
|:------------------------------------:|:------------------------------------:|:------------------------------------:|:------------------------------------:|
| ![](https://i.imgur.com/qeeMsVe.png) | ![](https://i.imgur.com/zruuelZ.png) | ![](https://i.imgur.com/y5mTUey.png) | ![](https://i.imgur.com/3wvwDVo.png) |

## To-do list

+ Add some utils to easier manage `BluetoothConnection` (see discussion #41),
+ Allow connection method/protocol/UUID specification,
+ Listening/server mode,
+ Recognizing and displaying `BluetoothClass` of device,
+ Maybe integration with `flutter_blue` one day ;)

You might also want to
check [milestones](https://github.com/edufolly/flutter_bluetooth_serial/milestones).

## Credits

- [Eduardo Folly](mailto:edufolly@gmail.com)
- [Martin Mauch](mailto:martin.mauch@gmail.com)
- [Patryk Ludwikowski](mailto:patryk.ludwikowski.7@gmail.com)

After version 0.3.0 we have a lot of collaborators. If you would like to be
credited, please send me an [email](mailto:edufolly@gmail.com).


CHANGELOG.md
## [0.4.0] - 2021-08-17

- Update flutter plugin v2.

## [0.3.2] - 2021-07-15

- Fixing pubspec.yaml typo.

## [0.3.1] - 2021-07-15

- Update example to sound null safety and new Material Buttons.
- Changed FutureOr<int> to FutureOr<dynamic> in getBondStateForAddress.
- Fix an unhandled exception when trying to get bluetooth state.

## [0.3.0] - 2021-07-04

- Implementing null safety.

## [0.2.2] - 2019-08-19

- Fix closing connections which are not `listen`ing to `input` (solved #60),
- Add more clear example of detecting which side caused disconnection,
- Add exception on `add`ing to `output` if connection closed,
- `BluetoothConnection` `cancel` is deprecated now, use `close` instead. It was
  stupid to name `cancel` something that `close`s (it was even documented that
  way, lol).

## [0.2.1] - 2019-08-05

- Apply patch #48 for some disconnection issues,
- Update `ChatPage` to newer API,
- Fix and update AndroidX declaration.

## [0.2.0] - 2019-07-02

- Bonding (Pairing):
    - starting outgoing pairing requests,
    - handling incoming pairing requests,
    - remove current bondings to devices,
- Discoverability!
    - requesting discoverable mode for specified duration,
    - example with timeout countdown,
    - checking discoverable mode.

And few more features:

- get/set for local (discoverable) device name,
- getting local adapter address (with some hacks to work on newer Androids),
- checking for `isConnected` for discovered or bonded devices,
- fixed few broadcast receiver leaks.

## [0.1.1] - 2019-07-01

- Patch #43 for "Error when meet unknown devices".

## [0.1.0] - 2019-06-19

Pull request #35 by @PsychoXIVI changes a lot:

- Discovering other devices,
- Multiple connections,
- Interesting example application,
- Enabling/Disabling Bluetooth,
- Byte streams,
- Overall improvements and fixes.

## [0.0.5] - 2019-03-18

- Upgrade for AndroidX support (thanks @akilinomendez)
- New default constructor (thanks @MohiuddinM)
- Added method write passing byte[] (thanks @rafaelterada)
- Upgrade to Android Studio 3.3.2

## [0.0.4] - 2018-12-20

- Unsupported value error correction (thanks @rafaelterada)
- Added openSettings, which opens the bluetooth settings screen (thanks
  @rafaelterada)

## [0.0.3] - 2018-11-22

- async connection and null exception handled (thanks @MohiuddinM)

## [0.0.2] - 2018-09-27

- isConnected Implementation (thanks @Riscue)

## [0.0.1] - 2018-08-20

- Only Android support.

lib/BluetoothConnection.dart
part of flutter_bluetooth_serial;

/// Represents ongoing Bluetooth connection to remote device.
class BluetoothConnection {
  // Note by PsychoX at 2019-08-19 while working on issue #60:
  // Fixed and then tested whole thing multiple times:
  // - [X] basic `connect`, use `output`/`input` and `disconnect` by local scenario,
  // - [X] fail on connecting to device that isn't listening,
  // - [X] working `output` if no `listen`ing to `input`,
  // - [X] closing by local if no `input` written,
  // - [X] closing by local if no `listen`ing to `input` (this was the #60),
  // - [X] closing by remote if no `input` written,
  // - [X] closing by remote if no `listen`ing to `input`,
  //    It works, but library user is notifed either by error on `output.add` or
  //    by observing `connection.isConnected`. In "normal" conditions user can
  //    listen to `input` even just for the `onDone` to proper detect closing.
  //

  /// This ID identifies real full `BluetoothConenction` object on platform side code.
  final int? _id;

  final EventChannel _readChannel;
  late StreamSubscription<Uint8List> _readStreamSubscription;
  late StreamController<Uint8List> _readStreamController;

  /// Stream sink used to read from the remote Bluetooth device
  ///
  /// `.onDone` could be used to detect when remote device closes the connection.
  ///
  /// You should use some encoding to receive string in your `.listen` callback, for example `ascii.decode(data)` or `utf8.encode(data)`.
  Stream<Uint8List>? input;

  /// Stream sink used to write to the remote Bluetooth device
  ///
  /// You should use some encoding to send string, for example `.add(ascii.encode('Hello!'))` or `.add(utf8.encode('Cześć!))`.
  late _BluetoothStreamSink<Uint8List> output;

  /// Describes is stream connected.
  bool get isConnected => output.isConnected;

  BluetoothConnection._consumeConnectionID(int? id)
      : this._id = id,
        this._readChannel =
            EventChannel('${FlutterBluetoothSerial.namespace}/read/$id') {
    _readStreamController = StreamController<Uint8List>();

    _readStreamSubscription =
        _readChannel.receiveBroadcastStream().cast<Uint8List>().listen(
              _readStreamController.add,
              onError: _readStreamController.addError,
              onDone: this.close,
            );

    input = _readStreamController.stream;
    output = _BluetoothStreamSink<Uint8List>(id);
  }

  /// Returns connection to given address.
  static Future<BluetoothConnection> toAddress(String? address) async {
    // Sorry for pseudo-factory, but `factory` keyword disallows `Future`.
    return BluetoothConnection._consumeConnectionID(await FlutterBluetoothSerial
        ._methodChannel
        .invokeMethod('connect', {"address": address}));
  }

  /// Should be called to make sure the connection is closed and resources are freed (sockets/channels).
  void dispose() {
    finish();
  }

  /// Closes connection (rather immediately), in result should also disconnect.
  Future<void> close() {
    return Future.wait([
      output.close(),
      _readStreamSubscription.cancel(),
      (!_readStreamController.isClosed)
          ? _readStreamController.close()
          : Future.value(/* Empty future */)
    ], eagerError: true);
  }

  /// Closes connection (rather immediately), in result should also disconnect.
  @Deprecated('Use `close` instead')
  Future<void> cancel() => this.close();

  /// Closes connection (rather gracefully), in result should also disconnect.
  Future<void> finish() async {
    await output.allSent;
    close();
  }
}

/// Helper class for sending responses.
class _BluetoothStreamSink<Uint8List> extends StreamSink<Uint8List> {
  final int? _id;

  /// Describes is stream connected.
  bool isConnected = true;

  /// Chain of features, the variable represents last of the futures.
  Future<void> _chainedFutures = Future.value(/* Empty future :F */);

  late Future<dynamic> _doneFuture;

  /// Exception to be returend from `done` Future, passed from `add` function or related.
  dynamic exception;

  _BluetoothStreamSink(this._id) {
    // `_doneFuture` must be initialized here because `close` must return the same future.
    // If it would be in `done` get body, it would result in creating new futures every call.
    _doneFuture = Future(() async {
      // @TODO ? is there any better way to do it? xD this below is weird af
      while (this.isConnected) {
        await Future.delayed(Duration(milliseconds: 111));
      }
      if (this.exception != null) {
        throw this.exception;
      }
    });
  }

  /// Adds raw bytes to the output sink.
  ///
  /// The data is sent almost immediately, but if you want to be sure,
  /// there is `this.allSent` that provides future which completes when
  /// all added data are sent.
  ///
  /// You should use some encoding to send string, for example `ascii.encode('Hello!')` or `utf8.encode('Cześć!)`.
  ///
  /// Might throw `StateError("Not connected!")` if not connected.
  @override
  void add(Uint8List data) {
    if (!isConnected) {
      throw StateError("Not connected!");
    }

    _chainedFutures = _chainedFutures.then((_) async {
      if (!isConnected) {
        throw StateError("Not connected!");
      }

      await FlutterBluetoothSerial._methodChannel
          .invokeMethod('write', {'id': _id, 'bytes': data});
    }).catchError((e) {
      this.exception = e;
      close();
    });
  }

  /// Unsupported - this ouput sink cannot pass errors to platfom code.
  @override
  void addError(Object error, [StackTrace? stackTrace]) {
    throw UnsupportedError(
        "BluetoothConnection output (response) sink cannot receive errors!");
  }

  @override
  Future addStream(Stream<Uint8List> stream) => Future(() async {
        // @TODO ??? `addStream`, "alternating simultaneous addition" problem (read below)
        // If `onDone` were called some time after last `add` to the stream (what is okay),
        // this `addStream` function might wait not for the last "own" addition to this sink,
        // but might wait for last addition at the moment of the `onDone`.
        // This can happen if user of the library would use another `add` related function
        // while `addStream` still in-going. We could do something about it, but this seems
        // not to be so necessary since `StreamSink` specifies that `addStream` should be
        // blocking for other forms of `add`ition on the sink.
        var completer = Completer();
        stream.listen(this.add).onDone(completer.complete);
        await completer.future;
        await _chainedFutures; // Wait last* `add` of the stream to be fulfilled
      });

  @override
  Future close() {
    isConnected = false;
    return this.done;
  }

  @override
  Future get done => _doneFuture;

  /// Returns a future which is completed when the sink sent all added data,
  /// instead of only if the sink got closed.
  ///
  /// Might fail with an error in case if some occured while sending the data.
  /// Typical error could be `StateError("Not connected!")` which could happen
  /// if disconnected in middle of sending (queued) `add`ed data.
  ///
  /// Otherwise, the returned future will complete when either:
  Future get allSent => Future(() async {
        // Simple `await` can't get job done here, because the `_chainedFutures` member
        // in one access time provides last Future, then `await`ing for it allows the library
        // user to add more futures on top of the waited-out Future.
        Future lastFuture;
        do {
          lastFuture = this._chainedFutures;
          await lastFuture;
        } while (lastFuture != this._chainedFutures);

        if (this.exception != null) {
          throw this.exception;
        }

        this._chainedFutures =
            Future.value(); // Just in case if Dart VM is retarded
      });
}
